File: project(2)/classes/Player.js
const X_VELOCITY = 100
const Y_VELOCITY = 100

class Player {
  constructor({ x, y, size, velocity = { x: 0, y: 0 } }) {
    this.x = x
    this.y = y
    this.width = size
    this.height = size
    this.velocity = velocity
    this.isImageLoaded = false
    this.image = new Image()
    this.image.onload = () => {
      this.isImageLoaded = true
    }
    this.image.src = './images/player.png'
    this.elapsedTime = 0
    this.currentFrame = 0
    this.sprites = {
      idle: {
        x: 0,
        y: 0,
        width: 16,
        height: 16,
        frames: 1,
      },
      walkDown: {
        x: 0,
        y: 0,
        width: 16,
        height: 16,
        frames: 4,
      },
      walkUp: {
        x: 16,
        y: 0,
        width: 16,
        height: 16,
        frames: 4,
      },
      walkLeft: {
        x: 32,
        y: 0,
        width: 16,
        height: 16,
        frames: 4,
      },
      walkRight: {
        x: 48,
        y: 0,
        width: 16,
        height: 16,
        frames: 4,
      },
    }
    this.currentSprite = this.sprites.idle
    this.facing = 'down'
  }

  draw(c) {
    // Red square debug code
    // c.fillStyle = 'rgba(0, 0, 255, 0.5)'
    // c.fillRect(this.x, this.y, this.width, this.height)

    if (this.isImageLoaded === true) {
      let xScale = 1
      let x = this.x
      const cropOffset = 0.5

      if (this.facing === 'left') {
        xScale = -1
        x = -this.x - this.width
      }

      c.save()
      if (this.isInvincible) {
        c.globalAlpha = 0.5
      } else {
        c.globalAlpha = 1
      }
      c.scale(xScale, 1)
      c.drawImage(
        this.image,
        this.currentSprite.x,
        this.currentSprite.y +
          this.currentSprite.height * this.currentFrame +
          cropOffset,
        this.currentSprite.width,
        this.currentSprite.height,
        x,
        this.y,
        this.width,
        this.height,
      )
      c.restore()
    }
  }

  update(deltaTime, collisionBlocks) {
    if (!deltaTime) return

    // Updating animation frames
    this.elapsedTime += deltaTime
    const secondsInterval = 0.1
    if (this.elapsedTime > secondsInterval) {
      this.currentFrame = (this.currentFrame + 1) % this.currentSprite.frames
      this.elapsedTime -= secondsInterval
    }

    // Update horizontal position and check collisions
    this.updateHorizontalPosition(deltaTime)
    this.checkForHorizontalCollisions(collisionBlocks)

    // Update vertical position and check collisions
    this.updateVerticalPosition(deltaTime)
    this.checkForVerticalCollisions(collisionBlocks)

    this.switchSprites()
  }

  switchSprites() {
    if (this.velocity.x === 0 && this.velocity.y === 0) {
      // Idle
      this.currentFrame = 0
      this.currentSprite.frames = 1
    } else if (
      this.velocity.x > 0 &&
      this.currentSprite !== this.sprites.walkRight
    ) {
      // Walk Right
      this.currentFrame = 0
      this.currentSprite = this.sprites.walkRight
      this.currentSprite.frames = 4
    } else if (
      this.velocity.x < 0 &&
      this.currentSprite !== this.sprites.walkLeft
    ) {
      // Walk Left
      this.currentFrame = 0
      this.currentSprite = this.sprites.walkLeft
      this.currentSprite.frames = 4
    } else if (
      this.velocity.y > 0 &&
      this.currentSprite !== this.sprites.walkDown
    ) {
      // Walk Down
      this.currentFrame = 0
      this.currentSprite = this.sprites.walkDown
      this.currentSprite.frames = 4
    } else if (
      this.velocity.y < 0 &&
      this.currentSprite !== this.sprites.walkUp
    ) {
      // Walk Up
      this.currentFrame = 0
      this.currentSprite = this.sprites.walkUp
      this.currentSprite.frames = 4
    }
  }

  updateHorizontalPosition(deltaTime) {
    this.x += this.velocity.x * deltaTime
  }

  updateVerticalPosition(deltaTime) {
    this.y += this.velocity.y * deltaTime
  }

  handleInput(keys) {
    this.velocity.x = 0
    this.velocity.y = 0

    if (keys.d.pressed) {
      this.velocity.x = X_VELOCITY
    } else if (keys.a.pressed) {
      this.velocity.x = -X_VELOCITY
    } else if (keys.w.pressed) {
      this.velocity.y = -Y_VELOCITY
    } else if (keys.s.pressed) {
      this.velocity.y = Y_VELOCITY
    }
  }

  checkForHorizontalCollisions(collisionBlocks) {
    const buffer = 0.0001
    for (let i = 0; i < collisionBlocks.length; i++) {
      const collisionBlock = collisionBlocks[i]

      // Check if a collision exists on all axes
      if (
        this.x <= collisionBlock.x + collisionBlock.width &&
        this.x + this.width >= collisionBlock.x &&
        this.y + this.height >= collisionBlock.y &&
        this.y <= collisionBlock.y + collisionBlock.height
      ) {
        // Check collision while player is going left
        if (this.velocity.x < -0) {
          this.x = collisionBlock.x + collisionBlock.width + buffer
          break
        }

        // Check collision while player is going right
        if (this.velocity.x > 0) {
          this.x = collisionBlock.x - this.width - buffer

          break
        }
      }
    }
  }

  checkForVerticalCollisions(collisionBlocks) {
    const buffer = 0.0001
    for (let i = 0; i < collisionBlocks.length; i++) {
      const collisionBlock = collisionBlocks[i]

      // If a collision exists
      if (
        this.x <= collisionBlock.x + collisionBlock.width &&
        this.x + this.width >= collisionBlock.x &&
        this.y + this.height >= collisionBlock.y &&
        this.y <= collisionBlock.y + collisionBlock.height
      ) {
        // Check collision while player is going up
        if (this.velocity.y < 0) {
          this.velocity.y = 0
          this.y = collisionBlock.y + collisionBlock.height + buffer
          break
        }

        // Check collision while player is going down
        if (this.velocity.y > 0) {
          this.velocity.y = 0
          this.y = collisionBlock.y - this.height - buffer
          break
        }
      }
    }
  }
}

File: project(2)/csharp/Player.cs
namespace BaconEggs;

public sealed class Player
{
    private const float XVelocity = 100f;
    private const float YVelocity = 100f;

    private float _elapsedTime;
    private int _currentFrame;
    private Sprite _currentSprite;

    public Player(float x, float y, float size)
    {
        X = x;
        Y = y;
        Width = size;
        Height = size;
        VelocityX = 0;
        VelocityY = 0;
        Sprites = new SpriteSet();
        _currentSprite = Sprites.Idle;
        Facing = "down";
        ImagePath = "./images/player.png";
        IsImageLoaded = true;
    }

    public float X { get; private set; }
    public float Y { get; private set; }
    public float Width { get; }
    public float Height { get; }
    public float VelocityX { get; private set; }
    public float VelocityY { get; private set; }
    public bool IsInvincible { get; set; }
    public bool IsImageLoaded { get; }
    public string ImagePath { get; }
    public string Facing { get; private set; }
    public SpriteSet Sprites { get; }

    public (float X, float Y) Center => (X + Width / 2f, Y + Height / 2f);

    public void Draw(IRenderContext context)
    {
        if (!IsImageLoaded)
        {
            return;
        }

        float xScale = 1;
        float drawX = X;
        const float cropOffset = 0.5f;

        if (Facing == "left")
        {
            xScale = -1;
            drawX = -X - Width;
        }

        context.Save();
        context.GlobalAlpha = IsInvincible ? 0.5f : 1f;
        context.Scale(xScale, 1);
        context.DrawImage(
            ImagePath,
            _currentSprite.X,
            _currentSprite.Y + _currentSprite.Height * _currentFrame + cropOffset,
            _currentSprite.Width,
            _currentSprite.Height,
            drawX,
            Y,
            Width,
            Height);
        context.Restore();
    }

    public void Update(float deltaTime, IReadOnlyList<CollisionBlock> collisionBlocks)
    {
        if (deltaTime <= 0)
        {
            return;
        }

        _elapsedTime += deltaTime;
        const float secondsInterval = 0.1f;
        if (_elapsedTime > secondsInterval)
        {
            _currentFrame = (_currentFrame + 1) % _currentSprite.Frames;
            _elapsedTime -= secondsInterval;
        }

        X += VelocityX * deltaTime;
        CheckForHorizontalCollisions(collisionBlocks);

        Y += VelocityY * deltaTime;
        CheckForVerticalCollisions(collisionBlocks);

        SwitchSprites();
    }

    public void HandleInput(InputState keys)
    {
        VelocityX = 0;
        VelocityY = 0;

        if (keys.D.Pressed)
        {
            VelocityX = XVelocity;
            Facing = "right";
        }
        else if (keys.A.Pressed)
        {
            VelocityX = -XVelocity;
            Facing = "left";
        }
        else if (keys.W.Pressed)
        {
            VelocityY = -YVelocity;
            Facing = "up";
        }
        else if (keys.S.Pressed)
        {
            VelocityY = YVelocity;
            Facing = "down";
        }
    }

    private void SwitchSprites()
    {
        if (VelocityX == 0 && VelocityY == 0)
        {
            _currentFrame = 0;
            _currentSprite = Sprites.Idle;
        }
        else if (VelocityX > 0 && _currentSprite != Sprites.WalkRight)
        {
            _currentFrame = 0;
            _currentSprite = Sprites.WalkRight;
        }
        else if (VelocityX < 0 && _currentSprite != Sprites.WalkLeft)
        {
            _currentFrame = 0;
            _currentSprite = Sprites.WalkLeft;
        }
        else if (VelocityY > 0 && _currentSprite != Sprites.WalkDown)
        {
            _currentFrame = 0;
            _currentSprite = Sprites.WalkDown;
        }
        else if (VelocityY < 0 && _currentSprite != Sprites.WalkUp)
        {
            _currentFrame = 0;
            _currentSprite = Sprites.WalkUp;
        }
    }

    private void CheckForHorizontalCollisions(IReadOnlyList<CollisionBlock> collisionBlocks)
    {
        const float buffer = 0.0001f;
        foreach (var block in collisionBlocks)
        {
            if (!Intersects(block))
            {
                continue;
            }

            if (VelocityX < 0)
            {
                X = block.X + block.Width + buffer;
                break;
            }

            if (VelocityX > 0)
            {
                X = block.X - Width - buffer;
                break;
            }
        }
    }

    private void CheckForVerticalCollisions(IReadOnlyList<CollisionBlock> collisionBlocks)
    {
        const float buffer = 0.0001f;
        foreach (var block in collisionBlocks)
        {
            if (!Intersects(block))
            {
                continue;
            }

            if (VelocityY < 0)
            {
                VelocityY = 0;
                Y = block.Y + block.Height + buffer;
                break;
            }

            if (VelocityY > 0)
            {
                VelocityY = 0;
                Y = block.Y - Height - buffer;
                break;
            }
        }
    }

    private bool Intersects(CollisionBlock block) =>
        X <= block.X + block.Width &&
        X + Width >= block.X &&
        Y + Height >= block.Y &&
        Y <= block.Y + block.Height;
}
